llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -1041,13 +1074,10 @@ findForkedPointer(PredicatedScalarEvolution &PSE,
   SmallVector<PointerIntPair<const SCEV *, 1, bool>> Scevs;
   findForkedSCEVs(SE, L, Ptr, Scevs, MaxForkedSCEVDepth);
 
-  // For now, we will only accept a forked pointer with two possible SCEVs
-  // that are either SCEVAddRecExprs or loop invariant.
+  // For now, we will only accept a forked pointer with two possible SCEVs.
   if (Scevs.size() == 2 &&
-      (isa<SCEVAddRecExpr>(get<0>(Scevs[0])) ||
-       SE->isLoopInvariant(get<0>(Scevs[0]), L)) &&
-      (isa<SCEVAddRecExpr>(get<0>(Scevs[1])) ||
-       SE->isLoopInvariant(get<0>(Scevs[1]), L))) {
+      hasComputableBounds(PSE, Ptr, get<0>(Scevs[0]), L, /*Assume*/ false) &&
```

Comment:
In this MR I've splitted preliminary NFC changes into the separate commit: https://github.com/llvm/llvm-project/pull/106574/commits/3415b7e7ea53a74a87bf7048b297971a67ed7ada; so this commit can moved into the separate PR if it will be easier to review (TBH I don't know what's the best practice of organizing stacked PRs on github)

Reply:
Created separate PR: https://github.com/llvm/llvm-project/pull/112571